modules/transactionpool/update.go

```diff
@@ -39,6 +46,46 @@ func (tp *TransactionPool) ProcessConsensusChange(cc modules.ConsensusChange) {
 	if err != nil {
 		tp.log.Println("ERROR: could not update the recent consensus change:", err)
 	}
+	err = tp.putBlockHeight(tp.dbTx, tp.blockHeight)
+	if err != nil {
+		tp.log.Println("ERROR: could not update the block height:", err)
+	}
+
+	// prune transactions older than maxTxnAge.
+	pruneTxns := make(map[types.TransactionID]struct{})
+	for txid, seenHeight := range tp.transactionHeights {
+		if tp.blockHeight-seenHeight > maxTxnAge {
+			delete(tp.transactionHeights, txid)
+			pruneTxns[txid] = struct{}{}
+		}
+	}
+	for id, tSet := range tp.transactionSets {
+		var validTxns []types.Transaction
+		for _, txn := range tSet {
+			_, shouldPrune := pruneTxns[txn.ID()]
+			if !shouldPrune {
+				validTxns = append(validTxns, txn)
+			}
+		}
+		tp.transactionSets[id] = validTxns
+		tp.transactionListSize -= len(encoding.Marshal(tSet)) - len(encoding.Marshal(validTxns))
+		if len(validTxns) == 0 {
+			delete(tp.transactionSets, id)
+			setDiffs, exists := tp.transactionSetDiffs[id]
+			if exists {
+				for _, diff := range setDiffs.SiacoinOutputDiffs {
+					delete(tp.knownObjects, ObjectID(diff.ID))
+				}
+				for _, diff := range setDiffs.FileContractDiffs {
+					delete(tp.knownObjects, ObjectID(diff.ID))
+				}
+				for _, diff := range setDiffs.SiafundOutputDiffs {
+					delete(tp.knownObjects, ObjectID(diff.ID))
+				}
+			}
+			delete(tp.transactionSetDiffs, id)
+		}
+	}
```

Comment:
all the transactions get thrown out at `t.purge()` already anyway, there's no need to manually extract them. Instead, after the purge happens, just scroll through the list of `unconfirmedSets` and check the transactions